### ISSUE TITLE
Browserless authentication?

### ISSUE THREAD

**Issue by Boereck** [deeplay-io/verdaccio-openid-connect#15]
Hi,

We are currently evaluating using the verdaccio-openid-connect plugin, however we do not want to rely on any browser interaction for logging in (not even the very first login), so this process can be automated as much as possible. 

It is unclear to me if the NPM client application is suited for such a direct workflow. We thought one chain of actions to implement this would be the following:
- Getting a bearer token from an OIDC server
- Authenticating at the registry with that bearer token
- Storing the auth-token returned by the registry in the `~/.npmrc` file

We could see a login mechanism where the bearer token is received externally (via a different tool) and passed to the npm adduser command as a base64 encoded string in the username.

We would also accept to have a tool external to NPM that performs all the actions described above if the verdaccio-openid-connect plugin would implement the workflow to turn a bearer token into an auth-token to be stored in the `~/.npmrc` file. We would also contribute to this project if we get some hints how to help to implement such a "browserless" login flow.

Best regards,
Max

**Comment by aikoven** [comment]
Hey,

With this plugin, the token you get in `.npmrc` is not the OIDC access token, but the token issued by Verdaccio. The OIDC tokens are stored by this plugin's persistence layer and are refreshed during requests. If we used OIDC token in `.npmrc` directly, users would have to constantly re-login because OIDC tokens have limited lifetime.

Like you say, we could introduce a new mode for the plugin, where it would accept OIDC tokens via the regular username + password flow in `npm login`. Note that the plugin needs not the `access_token` but `refresh_token` to keep the session alive. Also, I would rather use the password for passing token, not username.

**Comment by Boereck** [comment]
Hi,

Thanks for the quick reply! I would also not be in favor of storing the OIDC token in the `.npmrc` file, since we would need to keep the token fresh and this has to be done constantly outside of NPM.

So the "regular" login flow with username + password (where password = bearer token) would be a possiblity.  
We had a look at the plugin's source and played around a bit. It seems if the plugin does not respond on the `/-/v1/login` route, then the client will gather the login data from the user and call the `/-/user/org.couchdb.user:<username>` endpoint. The plugin can react on this with the `adduser` method, defined in the `IPluginAuth` interface. However, if `adduser` is implemented, verdaccio seems to take care of creating the auth token itself. There does not seem to be a way to force npm to directly call the user endpoint, so it seems the registry can only allow one way of logging in (either the browser based authentacation, or the discussed one using a bearer token). We can live with this, even though it would be nice to be able to allow both ways.

So the direction we would propose would be to add a configuration switch to enable the bearer token based login, and based on the switch either do or don't register a handler for the `/-/v1/login` route in `register_middlewares`. It seems another route has to be implemented for the route `/-/user/org.couchdb.user:<username>` to validate the bearer token, manage the session storage and token storage, and return the auth token to the user (similar to how the `/oidc/callback` route works). Does that seem to be a reasonable solution?

 If you would give some guidance on the implementation details, we could try to implement the flow and open a PR.

**Comment by Boereck** [comment]
We have done some experiments and found a workflow that works: As described above, we've added another route for `/-/user/org.couchdb.user:<username>`, which takes an access token as the password. To get a complete TokenSet, we then call the token endpoint of the OIDC server with the request to perform a token exchange operation to turn the access token into a refresh token. With the response we get an access token, refresh token, and with the parameter `scope=openid` also an id token. We can then use the same logic as in the route `/oidc/callback` to store the TokenSet and return the npmToken to the user. However, we only tested this with Keyclaok as the OIDC server.

If this sounds good to you, we can polish the code, add documentation and open a PR.
